bpaingest/sync.py

```diff
@@ -36,18 +44,36 @@ def sync_package(ckan, obj):
     return ckan_obj
 
 
+def build_package_cache(ckan, package_types):
+    """
+    build a cache of all the packages in `org`, to speed up comparison
+    """
+
+    packages = []
+    for typ in package_types:
+        logger.info("Retrieving all extant packages of type: {}".format(typ))
+        results = ckan_method(ckan, 'package', 'search')(q='type:{}'.format(typ), include_private=True, rows=50000)
+        packages += results['results']
+    logger.info("{} packages cached.".format(len(packages)))
+    return dict((t['id'], t) for t in packages)
+
+
 def sync_packages(ckan, packages, org, group):
     # FIXME: we don't check if there are any packages we should remove (unpublish)
     logger.info('syncing %d packages' % (len(packages)))
     # we have to post the group back in package objects, send a minimal version of it
     api_group_obj = prune_dict(group, ('display_name', 'description', 'title', 'image_display_url', 'id', 'name'))
     ckan_packages = []
+
+    cache = build_package_cache(ckan, set(t['type'] for t in packages))
```

Comment:
I would pass in just packages here and making sure that the types are unique would be the concern of the `build_package_cache` function not of the client code's invoking it.